src/managed/Microsoft.DotNet.Build.Bundle/BundleManifest.cs

```diff
@@ -0,0 +1,208 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.DotNet.Build.Bundle
+{
+    /// <summary>
+    ///  BundleManifest is a description of the contents of a bundle file.
+    ///  This class handles creation and consumption of bundle-manifests.
+    ///  
+    ///  Here is the description of the Bundle Layout:
+    ///  _______________________________________________
+    ///  AppHost 
+    ///
+    ///
+    /// ----------------------------------------------
+    /// The embedded files including the app, its
+    /// configuration files, dependencies, and 
+    /// possibly the runtime.
+    /// 
+    /// 
+    /// 
+    /// 
+    /// 
+    /// 
+    ///
+    /// -----------------------------------------------
+    /// Meta-data Directory (FileEntries)
+    ///   "app.deps.json"               (if any)
+    ///   "app.runtimeconfig.json"      (if any)
+    ///   "app.runtimeconfig.dev.json"  (if any)
+    ///   "app.dll"                 (The main app)
+    ///   "IL_1.dll"                (Assemblies, if any)
+    ///    ...
+    ///   "IL_n.dll"
+    ///   "File_1"                  (Other Files, if any)
+    ///    ...
+    ///   "File_n" 
+    /// - - - - - - - - - - - - - - - - - - - - - - - -
+    ///  Bundle Manifest (footer)
+    ///     MajorVersion
+    ///     MinorVersion
+    ///     FileEntryStart (offset)
+    ///     NumFiles
+    ///     NumAssemblies
+    ///     Flags
+    /// -----------------------------------------------
+    ///   Bundle Signature
+    /// _________________________________________________
+    /// 
+    /// The Bundle manifest does not encode FileType explicitly in each entry. 
+    /// The FileType for each embedded file is deduced based on:
```

Comment:
This works. My preference would be to instead store type for each entry. There's no added startup cost as the host will have to parse all the entries anyway to spill files onto disk. And it leaves the format a bit more flexible - as is we enforce certain ordering for what feels like arbitrary reason. But it's just my personal preference, feel free to ignore it.